scarb/src/bin/scarb/commands/commands.rs

```diff
@@ -1,9 +1,159 @@
 use anyhow::Result;
+use serde::{Serialize, Serializer};
+use std::collections::BTreeMap;
+use std::fmt::Display;
+use std::path::PathBuf;
+use std::{env, fs};
 
+use crate::args::ScarbArgs;
 use scarb::core::Config;
+use scarb::process::is_executable;
+use scarb::ui::Message;
+use scarb::EXTERNAL_CMD_PREFIX;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Serialize)]
+pub enum CommandInfo {
+    BuiltIn { about: Option<String> },
+    External { path: PathBuf },
+}
+
+impl Display for CommandInfo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CommandInfo::BuiltIn { about: Some(about) } => write!(f, "{}", about),
+            CommandInfo::BuiltIn { about: None } => write!(f, "",),
+            CommandInfo::External { path } => write!(f, "{}", path.display()),
+        }
+    }
+}
+
+impl Message for CommandInfo {
+    fn text(self) -> String {
+        self.to_string()
+    }
+
+    fn structured<S: Serializer>(self, ser: S) -> Result<S::Ok, S::Error> {
+        self.serialize(ser)
+    }
+}
+
+#[derive(Serialize, Debug)]
+pub struct CommandsList {
+    pub commands: BTreeMap<String, CommandInfo>,
+}
+
+impl Message for CommandsList {
+    fn text(self) -> String {
+        let mut text = String::from("Installed Commands:\n");
+        for (name, info) in self.commands.iter() {
+            text.push_str(&format!("{:<20}: {}\n", name, info.to_string()));
+        }
+        text
+    }
+
+    fn structured<S: Serializer>(self, ser: S) -> Result<S::Ok, S::Error> {
+        self.commands.serialize(ser)
+    }
+}
+
+#[tracing::instrument(level = "debug", skip(config, builtins))]
+pub fn list_commands(
```

Comment:
```suggestion
fn list_commands(
```

1. no need to instrument this because this is wrapped by `run` pretty swiftly
2. no need to pub it either because this is going only to be used here I guess?